driver-core/src/main/com/mongodb/internal/connection/EventHelper.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.internal.connection;
+
+import com.mongodb.connection.ClusterDescription;
+import com.mongodb.connection.ServerDescription;
+
+import java.util.Objects;
+
+final class EventHelper {
+
+    static boolean equivalentEvent(final ClusterDescription current, final ClusterDescription previous) {
+        if (!exceptionsEquals(current.getSrvResolutionException(), previous.getSrvResolutionException())) {
+            return false;
+        }
+        if (current.getServerDescriptions().size() != previous.getServerDescriptions().size()) {
+            return false;
+        }
+        for (ServerDescription curNew: current.getServerDescriptions()) {
+            ServerDescription matchingPrev = null;
+            for (ServerDescription curPrev: previous.getServerDescriptions()) {
+                if (curNew.getAddress().equals(curPrev.getAddress())) {
+                    matchingPrev = curPrev;
+                    break;
+                }
+            }
+            if (!equivalentEvent(curNew, matchingPrev)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    static boolean equivalentEvent(final ServerDescription current, final ServerDescription previous) {
```

Comment:
The logic here looks almost identical to `ServerDescription#equals`, just in a different order of the checks. Is there a reason this is duplicated? Or am I potentially missing something?

Reply:
It is similar, but it checks fewer fields (according to the rule of equality from the spec, which is not what is implemented in `ServerDescription#equals`.  It didn't seem worth pulling out the common bits and sharing them, since it's pretty much boilerplate code.  I considered putting this code in a public ServerDescription method, but didn't think it was necessary to commit to a public API for this logic.